### PR TITLE
Changes to throwing

### LINKED ISSUE #3824: Banana cream pies disappear when caught
Keep throw intent on, have someone throw a pie at you, normally this would make you catch it but it just disappears.

### LINKED ISSUE #6569: Lost momentum after exiting a mass drivered container mid-flight
If you get shot out of a mass driver in an object, say the coffin on the mass driver in the derelict, and unknowingly open it mid-flight for fun, it'll dump you out in the middle of space and the coffin will keep traveling by itself. It's not a bug, it's just unintuitive behaviour.

### LINKED ISSUE #10428: Alien hunter leaping shifts a tile to the left upon smashing into something.
I jumped straight up in this scenario, and shifted a tile to the left upon smashing into the wall for a split second, then back to the appropriate tile.

### LINKED ISSUE #10479: Thrown objects cannot pass through a border window tile
This has been the case since the dawn of time.

If you try to throw an object through the tile of a border window, perpendicular to the window in question, it will hit the window. This extends to all border objects like this such as border firelocks, and alien hunters experience this since I'm told they use something similar if not identical to being thrown.

### LINKED ISSUE #10507: Crusher still kills when off
The crusher was off, the panel was open and it was unsecured. I was dragging it to a safe place. Someone accidentally walked into it, and was immediately killed by it. All their possessions were destroyed. This happened twice. It was hilarious.

### LINKED ISSUE #10536: Catching thrown items is broken
How to reproduce:

1. Try to catch something by enabling throwing mode.
2. It hits you instead.

### PR DESCRIPTION
**Pull request by phil235**
- Fixed catching banana creampie in mid throw (and similar objs). Fixes #3824 , Fixes #10536 , (hitting a wall now makes the smashed pie effect appear ON the wall)
- Some additional adjustments to toy/balloon.
- Moved last_bumped var from movable to living.
- Renamed /obj/machinery/field/proc/bump to bump_field() to avoid confusion with Bump(). Same thing with /obj/effect/mine/Bumped() -> triggermine().
- Fixes sprite when hunter leaps into a wall. Fixes #10428
- Removed some commented code here and there.
- Remove allow_spin var, throw_at now uses a spin argument for that. 
- Replaced the "throwpass" atom var by the pass flag LETPASSTHROW.
- A thrown mob hitting another mob now produces a sound; also thrown mob and target one no longer swap places even on help intent.
- A thrown mob now is lightly hurt (and weakened) if he hits a wall/mob/dense object. I Nerfed the damage when thrown mob hit wall.(20 -> 10)
- Thrown obj/mob no longer bounces off wall unless it's no grav. Heavy thrown items now push an unanchored obj/mob target.
- Fixes losing all momentum when getting out of a thrown closet. Fixes #6569
- A lot of work on throw_at, throw_impact, and hitby to make the code more OOP. Thrown items no longer collide with border items on the side. Fixes #10479. The default throw_at() now uses diagonal directions. 
- Fixes the killer crusher. Fixes #10507 
- Removes useless gateway/Crossed and gateway/Bumped().
- I added the item var "thrownby" to be able to continue to log the thrower of the item when it hits a mob. It removes the need for a thrower argument in throw_impact() and hitby().

### CODE REVIEW

File: code/modules/power/singularity/singularity.dm [hunk 1]
```diff
@@ -26,7 +26,7 @@
 	var/last_failed_movement = 0//Will not move in the same dir if it couldnt before, will help with the getting stuck on fields thing
 	var/last_warning
 	var/consumedSupermatter = 0 //If the singularity has eaten a supermatter shard and can go to stage six
-	allow_spin = 0
```
Comment (Iamgoofball) [inline code/modules/power/singularity/singularity.dm]: hey hey hey

this creates lag

File: code/game/atoms_movable.dm [hunk 2]
```diff
@@ -154,57 +149,60 @@
...
+	return ..()
+
+/atom/movable/proc/throw_at(atom/target, range, speed, mob/thrower, spin=1)
 	if(!target || !src || (flags & NODROP))	return 0
 	//use a modified version of Bresenham's algorithm to get from the atom's current position to that of the target
 
-	src.throwing = 1
-	if(target.allow_spin) // turns out 1000+ spinning objects being thrown at the singularity creates lag - Iamgoofball
+	throwing = 1
+	if(spin) // turns out 1000+ spinning objects being thrown at the singularity creates lag - Iamgoofball
```
Comment (Iamgoofball) [inline code/game/atoms_movable.dm]: Where is this defined?

Comment (RemieRichards) [inline code/game/atoms_movable.dm]: the proc args dummy.

**Comment by RemieRichards** [comment]
@phil235 does the new throw_at() correctly use diagonal dirs? it was a key reason I wrote the horrors of the xenomorph override of throw_at(). if it does not I have to give that aspect a :-1: as it looks HORRIFIC without it.

**Comment by phil235** [comment]
Yeah, I fucked up the xeno throw_at(), I'll fix it.

File: code/game/atoms_movable.dm [hunk 3]
```diff
@@ -93,17 +93,13 @@
...
+/atom/movable/Bump(var/atom/A as mob|obj|turf|area, yes) //the "yes" arg is to differentiate our Bump proc from byond's, without it every Bump() call would become a double Bump().
+	if((A && yes))
+		if(throwing)
+			throwing = 0
+			throw_impact(A)
+			. = 1
 		A.Bumped(src)
-	return
-	..()
-	return
```
Comment (MrPerson) [inline code/game/atoms_movable.dm]: rip in peace

On a more serious note, this shouldn't have the as mob|obj|turf|area bit.

File: code/game/gamemodes/cult/cult_structures.dm [hunk 4]
```diff
@@ -46,11 +35,7 @@
 	anchored = 1.0
 
 /obj/effect/gateway/Bumped(mob/M as mob|obj)
-	spawn(0)
-		return
```
Comment (MrPerson) [inline code/game/gamemodes/cult/cult_structures.dm]: why.jpg
Actually why is this proc even here? Seems superfluous at this point.

File: code/game/atoms_movable.dm (hunk 3 above)
Comment (phil235) [inline code/game/atoms_movable.dm]: I will remove it.

File: code/game/objects/structures/crates_lockers/closets.dm [hunk 5]
```diff
@@ -82,12 +82,18 @@
...
+				step(O, dir)
 
 	for(var/mob/M in src)
 		M.loc = loc
 		if(M.client)
 			M.client.eye = M.client.mob
 			M.client.perspective = MOB_PERSPECTIVE
+		if(throwing)
+			spawn(2)
+				step(M, dir)
```
Comment (MrPerson) [inline code/game/objects/structures/crates_lockers/closets.dm]: But shouldn't the closet be open and thus non-dense at this point? How on earth would anything collide with it?

File: code/game/gamemodes/cult/cult_structures.dm (hunk 4 above)
Comment (phil235) [inline code/game/gamemodes/cult/cult_structures.dm]: I thought this was here to override obj/Bumped() or effect/Bumped() or something. But apparently there's no non-empty parent Bumped(), so yeah let's remove.

File: code/game/objects/structures/crates_lockers/closets.dm (hunk 5 above)
Comment (phil235) [inline code/game/objects/structures/crates_lockers/closets.dm]: The mob comes out of the closet, makes one step forward, then the open closet (still thrown) moves forward and bumps into the mob. The mob is hit by it and the closet now moves backwards.

**Comment by phil235** [comment]
It should be ready for review.

**Comment by bustygoku** [comment]
Conflict

**Comment by phil235** [comment]
fixed

**Pull request merged**